lib/c-api/src/wasm_c_api/unstable/middlewares/metering.rs

```diff
@@ -0,0 +1,300 @@
+//!     wasm_config_push_middleware(config, middleware);
+//!     
+//!     // Create the engine and the store based on the configuration.
+//!     wasm_engine_t* engine = wasm_engine_new_with_config(config);
+//!     wasm_store_t* store = wasm_store_new(engine);
+//!     
+//!     // Create the new WebAssembly module.
+//!     wasm_byte_vec_t wat;
+//!     wasmer_byte_vec_new_from_string(
+//!         &wat,
+//!         "(module\n"
+//!         "  (type $add_t (func (param i32) (result i32)))\n"
+//!         "  (func $add_two_f (type $add_t) (param $value i32) (result i32)\n"
+//!         "    local.get $value\n"
+//!         "    i32.const 1\n"
+//!         "    i32.add)\n"
+//!         "  (export \"add_two\" (func $add_two_f)))"
+//!     );
+//!     wasm_byte_vec_t wasm;
+//!     wat2wasm(&wat, &wasm);
+//!
+//!     wasm_module_t* module = wasm_module_new(store, &wasm);
+//!     assert(module);
+//!     
+//!     // Instantiate the module.
+//!     wasm_extern_vec_t imports = WASM_EMPTY_VEC;
+//!     wasm_trap_t* traps = NULL;
+//!     wasm_instance_t* instance = wasm_instance_new(store, module, &imports, &traps);
+//!     assert(instance);
+//!     
+//!     // Here we go. At this step, we will get the `add_two` exported function, and
+//!     // call it.
+//!     wasm_extern_vec_t exports;
+//!     wasm_instance_exports(instance, &exports);
+//!     assert(exports.size >= 1);
+//!     assert(wasm_extern_kind(exports.data[0]) == WASM_EXTERN_FUNC);
+//!
+//!     const wasm_func_t* add_two = wasm_extern_as_func(exports.data[0]);
+//!     assert(add_two);
+//!
+//!     wasm_val_t arguments[1] = { WASM_I32_VAL(41) };
+//!     wasm_val_t results[1] = { WASM_INIT_VAL };
+//!
+//!     wasm_val_vec_t arguments_as_array = WASM_ARRAY_VEC(arguments);
+//!     wasm_val_vec_t results_as_array = WASM_ARRAY_VEC(results);
+//!
+//!     // Let's call `add_two` for the first time!
+//!     {
+//!         wasm_trap_t* trap = wasm_func_call(add_two, &arguments_as_array, &results_as_array);
+//!         assert(trap == NULL);
+//!         assert(results[0].of.i32 == 42);
+//!
+//!         // There is 6 points left!
+//!         assert(wasmer_metering_get_remaining_points(instance) == 6);
+//!     }
+//!
+//!     // Let's call `add_two` for the second time!
+//!     {
+//!         wasm_trap_t* trap = wasm_func_call(add_two, &arguments_as_array, &results_as_array);
+//!         assert(trap == NULL);
+//!         assert(results[0].of.i32 == 42);
+//!
+//!         // There is 2 points left!
+//!         assert(wasmer_metering_get_remaining_points(instance) == 2);
+//!     }
+//!
+//!     // Let's call `add_two` for the third time!
+//!     {
+//!         wasm_trap_t* trap = wasm_func_call(add_two, &arguments_as_array, &results_as_array);
+//!         // Oh, it failed!
+//!         assert(trap != NULL);
+//!
+//!         // There is 0 point left… they are exhausted.
+//!         assert(wasmer_metering_get_remaining_points(instance) == 0);
```

Comment:
Using `0` as the exhausted value may be unclear: did we only consume all the points we had? did we consume more than what we had before?

In the first case, it's not an error AFAIR: we consumed everything but not more than what we had. In the latter case, it's an error, we tried to consume more points than we had.

I think, we should either:
* have `-1` to tell we did not had enough points
* reintroduce a second function to tell if we exhausted or not.

Reply:
Hmm I see. I don't want to return -1 because we are dealing with `u64` already. Supporting negative value will reduce the set of values we can return. Maybe we could continue to return 0, and add a function like `wasmer_metering_points_is_exhausted(wasm_instance_t*)`. Thoughts?

Reply:
I kind of prefer using -1. I'm not sure the metering points will ever reach super ultra big values (specially in blockchain contract context).

Reply:
After discussion with @nlewycky, we decided to return `u64::MAX` so that we don't lose any values from `u64` by returning -1. I've also added `wasmer_metering_points_are_exhausted` to differentiate between 0 or exhausted points. Tests are up-to-date.